x-pack/plugins/security_solution/public/common/components/import_data_modal/utils.ts

```diff
@@ -0,0 +1,93 @@
+
+import * as i18n from './translations';
+
+import {
+  ImportDataResponse,
+  ImportResponseError,
+  ImportRulesResponseError,
+  ExceptionsImportError,
+} from '../../../detections/containers/detection_engine/rules';
+
+export const formatError = (
+  i18nFailedDetailedMessage: (message: string) => string,
+  importResponse: ImportDataResponse,
+  errors: Array<ImportRulesResponseError | ImportResponseError | ExceptionsImportError>
+) => {
+  const formattedErrors = errors.map((e) => i18nFailedDetailedMessage(e.error.message));
+  const error: Error & { raw_network_error?: object } = new Error(formattedErrors.join('. '));
+  error.stack = undefined;
+  error.name = 'Network errors';
+  error.raw_network_error = importResponse;
+
+  return error;
+};
+
+export const showToasterMessage = ({
+  importResponse,
+  exceptionsIncluded,
+  successMessage,
+  errorMessage,
+  errorMessageDetailed,
+  addError,
+  addSuccess,
+}: {
+  importResponse: ImportDataResponse;
+  exceptionsIncluded: boolean;
+  successMessage: (totalCount: number) => string;
+  errorMessage: (totalCount: number) => string;
+  errorMessageDetailed: (message: string) => string;
+  addError: (error: unknown, options: ErrorToastOptions) => Toast;
+  addSuccess: (toastOrTitle: ToastInput, options?: ToastOptions | undefined) => Toast;
+}) => {
+  // if import includes exceptions
+  if (exceptionsIncluded) {
+    // rules response actions
+    if (importResponse.success && importResponse.success_count > 0) {
+      addSuccess(successMessage(importResponse.success_count));
+    }
+
+    if (importResponse.errors.length > 0 && importResponse.rules_count) {
+      const error = formatError(errorMessageDetailed, importResponse, importResponse.errors);
+      addError(error, {
+        title: errorMessage(importResponse.rules_count - importResponse.success_count),
+      });
+    }
+
+    // exceptions response actions
+    if (
+      importResponse.exceptions_success &&
+      importResponse.exceptions_success_count != null &&
+      importResponse.exceptions_success_count > 0
+    ) {
+      addSuccess(i18n.SUCCESSFULLY_IMPORTED_EXCEPTIONS(importResponse.exceptions_success_count));
+    }
+
+    if (importResponse.exceptions_errors != null && importResponse.exceptions_errors.length > 0) {
+      const error = formatError(
+        errorMessageDetailed,
+        importResponse,
+        importResponse.exceptions_errors
+      );
+
+      addError(error, { title: i18n.IMPORT_FAILED(importResponse.exceptions_errors.length) });
+    }
+  } else {
+    // rules response actions
+    if (importResponse.success) {
+      addSuccess(successMessage(importResponse.success_count));
+    }
+
+    if (importResponse.errors.length > 0) {
+      const error = formatError(errorMessageDetailed, importResponse, importResponse.errors);
+      addError(error, { title: errorMessage(importResponse.errors.length) });
+    }
+  }
```

Comment:
Just a side note, no action is needed in this PR. 

The logic of this method seems pretty complicated to me. I think it is because we use the `ImportDataModal` component for two completely different scenarios: 1) Import of rules and 2) import of timelines. This logic could become even more challenging to follow, and we could reach a combinatorial explosion at some point if we continue reusing that component. So I think we should split it into smaller reusable components with business and UI logic separated.

Reply:
++ totally agree. Our UI has evolved a lot since this component was first created. It's def worth splitting next round of changes.